### ISSUE TITLE
VM Data Source

### ISSUE THREAD

**Issue by sei-aschlackman** [bpg/terraform-provider-proxmox#261]
**Is your feature request related to a problem? Please describe.**
I need to be able to look up a virtual machine resource. Specifically, I need to lookup a virtual machine template by tag to then use as the clone source of a new virtual machine.

**Describe the solution you'd like**
A VM Data Source

I would like to create a PR to add a new VM Data Source to the provider that would allow the lookup of virtual machines and templates by id or tag. I want to make sure that is something that you would be ok with before starting to work on it. 

I did see https://github.com/bpg/terraform-provider-proxmox/issues/29, in which you said you would look into creating such a VM Data Source, but it looks like nothing came of that. Were there any issues that blocked it or was it just not a priority?

Thanks for the great provider, and please let me know if there is anything I should do or watch out for to ensure that this feature can be merged successfully.

**Comment by bpg** [comment]
Hey @sei-aschlackman! 👋🏼

VM datasource wasn't a major priority for me, and no much votes / comments were given to #29 so it was eventually closed as stale.

The VM has tons of attributes. Although I can probably reuse quite a bit of code from the VM resource implementation, testing and validating the correctness of the imported data in the datasource is a daunting task. However, I believe we can create a "bare minimum" data source that only includes the attributes you're particularly interested in. Could you please provide a more concrete example of how you envision using it in your configuration? Once we have that, we can discuss implementation details and timelines.

I'd also like to mention that I don't want to add any major new functionality until #246 is completed. That code includes significant refactorings that I want to merge first, before adding anything else, to avoid major conflicts.

**Comment by sei-aschlackman** [comment]
A bare minimum data source would be fine. I really only need the id, and maybe the name. 

My main use case is for a terraform module that takes a tag as an input, finds a vm template with that tag and clones it. I need to have a stable unique identifier that users of these modules can reference and the id doesn't work for my use case because the modules will be used across multiple Proxmox installations and the same template may have a different integer id in each environment. 

Thanks for your quick response. I'm happy to wait for you to finish #246. Do you have any ETA on that? I could also fork from that branch if it's close to finished and then reconcile the final changes before making my PR.

**Comment by bpg** [comment]
That feature takes longer than I initially expected. Thought I may be able to push the refactored code before completing the whole thing, probably some time this weekend.

**Comment by bpg** [comment]
Hey @sei-aschlackman, I didn't have enough time to finish what I've planned with the refactoring and merge the code, so decided to go with "plan B" and just add the datasource to the current code as it is now.
As we discussed, it is very minimalistic. Pls. take a look at #268, and give it a try if you can.

**Comment by sei-aschlackman** [comment]
Awesome, thanks for putting that together. I was able to compile it locally and give it a try and it works as documented to lookup a vm by it's id or multiple vms by node, though I still need the ability to look up a vm by it's tag. So it would be something like

```
data "proxmox_virtual_environment_vm" "example" {
  tag = 'my_example_tag'
}
```

Which would try to find a vm with that tag on any node in the cluster, or

```
data "proxmox_virtual_environment_vm" "example" {
  tag = 'my_example_tag'
  node_name = 'pve'
}
```

to find a vm with that tag on a specific node. Ideally, it would throw an error if multiple vms with that tag were found. You would use the `proxmox_virtual_environment_vms` if you are ok with multiple results.

The Proxmox API doesn't return tags when getting all vm resources in a cluster, so I think it would have to list all nodes and then call get all vms once per node, which is kind of awkward but would be very useful for me. I'll try to see if I can get add that to what you wrote if you don't get a chance to. Thanks again.

**Comment by bpg** [comment]
Thanks, I've incorporated your feedback in the `proxmox_virtual_environment_vms` datasource.

I respectfully disagree with the behavior you've described for `proxmox_virtual_environment_vm` though :). In my opinion, it would be more appropriate for a singular VM datasource to be able to match only one concrete provider's resource, and to have a stable identifier associated with that resource.

Let me know what do you think.

**Comment by sei-aschlackman** [comment]
Thanks for the great work! I completely understand your concern about the `proxmox_virtual_environment_vm`. The only reason I'm trying to mess with tags in the first place is because Proxmox does **not** have a stable identifier for a vm template that spans across different Proxmox environments. If they used a uuid/guid instead of an int, it would be much simpler. 

The changes you made to `proxmox_virtual_environment_vms` will give me exactly what I need though. I tested it out and can achieve my desired result with the help of a terraform postcondition, like:

```
data "proxmox_virtual_environment_vms" "example" {
  id = "example"
  tags = ["38375a49-96a7-467b-a5f9-0d72ca28cb3b"]

  lifecycle {
    postcondition {
      condition     = length(self.vms) == 1
      error_message = "Only 1 vm should have this tag"
    }
  }
}
``` 

That way I will get an error if there is not exactly 1 vm with the specified tag. This will help me out greatly.

I do have one small suggestion though, regarding the id parameter on `proxmox_virtual_environment_vms`. It seems unnecessary for the user to have to specify an id here since it doesn't relate to anything in Proxmox. In a similar situation in the official vsphere provider, they use [time.Now().UTC().String()](https://github.com/hashicorp/terraform-provider-vsphere/blob/main/vsphere/data_source_vsphere_vmfs_disks.go#L73). Another alternative might be to generate a uuid/guid with something like https://github.com/google/uuid. It's not a big deal either way though and it works great as is. 

Thanks again!

**Comment by bpg** [comment]
You're welcome! And thank you for the feedback, much appreciated!
I was going back and forth with the dynamic `id` vs supplied `id`, but if the vsphere provider is doing that then we should be good too 😄
I'll make that change along with some other cleanups and merge the PR soon.
